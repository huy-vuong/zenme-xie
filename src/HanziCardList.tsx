import PropTypes from 'prop-types';
import React from 'react';
import HanziCard from 'zenme-xie/HanziCard';

export default function HanziCardList({ text }: HanziCardListProp) {
  return (
    <div>
      {text.split('').map((character, index) => (
        <div
          data-testid={`hanzi-card-${index}-${character.charCodeAt(0)}`}
          key={`hanzi-card-${index}-${character.charCodeAt(0)}`}
        >
          <HanziCard character={character} index={index} />
        </div>
      ))}
    </div>
  );
}

interface HanziCardListProp {
  text: string;
}

HanziCardList.propTypes = {
  text: PropTypes.string.isRequired,
};
