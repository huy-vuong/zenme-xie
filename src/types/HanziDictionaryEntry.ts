export default interface CharacterData {
  character: string;
  definition: string;
  pinyin: Array<string>;
  matches: Array<Array<number>>;
}
