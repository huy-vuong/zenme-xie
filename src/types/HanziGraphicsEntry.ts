export default interface HanziGraphicsEntry {
  strokes: Array<string>;
  medians: Array<Array<Array<number>>>;
  radStrokes: Array<number>;
}
