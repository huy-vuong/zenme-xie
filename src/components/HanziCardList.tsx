import PropTypes from 'prop-types';
import React from 'react';
import HanziCard from 'zenme-xie/components/HanziCard';
import HanziDictionaryEntry from 'zenme-xie/types/HanziDictionaryEntry';

export default function HanziCardList({ text, dictionary }: HanziCardListProp) {
  return (
    <>
      {text.split('').map((character, index) => (
        <div
          data-testid={`hanzi-card-${index}-${character.charCodeAt(0)}`}
          key={`hanzi-card-${index}-${character.charCodeAt(0)}`}
        >
          <HanziCard
            character={character}
            dictionaryEntry={dictionary.get(character) ?? null}
            index={index}
          />
        </div>
      ))}
    </>
  );
}

interface HanziCardListProp {
  text: string;
  dictionary: Map<string, HanziDictionaryEntry | null>;
}

HanziCardList.propTypes = {
  text: PropTypes.string.isRequired,
  dictionary: PropTypes.objectOf(
    PropTypes.shape({
      character: PropTypes.string,
      definition: PropTypes.string,
      pinyin: PropTypes.arrayOf(PropTypes.string),
    })
  ),
};
