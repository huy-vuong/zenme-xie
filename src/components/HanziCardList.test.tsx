import { render, waitFor } from '@testing-library/react';
import React from 'react';
import HanziCardList from 'zenme-xie/components/HanziCardList';
import HanziDictionaryEntry from 'zenme-xie/types/HanziDictionaryEntry';
import HanziGraphicsEntry from 'zenme-xie/types/HanziGraphicsEntry';

test('renders nothing given empty text', () => {
  const { asFragment } = render(
    <HanziCardList text="" dictionary={new Map()} graphics={new Map()} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders cards for each of the given characters', () => {
  const { getByTestId } = render(
    <HanziCardList
      text="貓咪喵喵叫"
      dictionary={dictionary}
      graphics={graphics}
    />
  );
  waitFor(() => {
    getByTestId('hanzi-card-0-35987');
    getByTestId('hanzi-card-1-21674');
    getByTestId('hanzi-card-2-21941');
    getByTestId('hanzi-card-3-21941');
    getByTestId('hanzi-card-4-21483');
  });
});

test('renders cards even for non-Hanzi characters', () => {
  const { getByTestId } = render(
    <HanziCardList text="Nyan?" dictionary={new Map()} graphics={new Map()} />
  );
  waitFor(() => {
    getByTestId('hanzi-card-0-78');
    getByTestId('hanzi-card-1-121');
    getByTestId('hanzi-card-2-97');
    getByTestId('hanzi-card-3-110');
    getByTestId('hanzi-card-4-63');
  });
});

const dictionary = new Map<string, HanziDictionaryEntry>([
  [
    '貓',
    {
      character: '貓',
      definition: 'cat',
      pinyin: ['māo'],
      matches: [
        [0],
        [0],
        [0],
        [0],
        [0],
        [0],
        [0],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1],
      ],
    },
  ],
  [
    '咪',
    {
      character: '咪',
      definition: "a cat's meow; a meter",
      pinyin: ['mī'],
      matches: [[0], [0], [0], [1], [1], [1], [1], [1], [1]],
    },
  ],
  [
    '喵',
    {
      character: '喵',
      definition: "a cat's meow",
      pinyin: ['miāo'],
      matches: [[0], [0], [0], [1], [1], [1], [1], [1], [1], [1], [1]],
    },
  ],
  [
    '叫',
    {
      character: '叫',
      definition: 'cry, shout; to call, to greet, to hail',
      pinyin: ['jiào'],
      matches: [[0], [0], [0], [1], [1]],
    },
  ],
]);
const graphics = new Map<string, HanziGraphicsEntry>([
  [
    '貓',
    {
      character: '貓',
      strokes: [
        'M 285 759 Q 246 708 139 631 Q 135 622 142 620 Q 145 621 155 626 Q 230 653 311 712 Q 348 736 383 750 Q 398 754 400 761 Q 401 771 389 780 Q 367 790 324 798 Q 308 802 297 798 Q 291 794 292 784 Q 295 771 285 759 Z',
        'M 128 546 Q 138 530 149 509 Q 156 496 170 493 Q 179 492 185 501 Q 192 511 191 537 Q 188 558 133 585 Q 123 591 116 590 Q 112 589 110 578 Q 111 568 128 546 Z',
        'M 227 632 Q 263 572 280 568 Q 290 565 298 579 Q 301 589 296 608 Q 281 639 235 653 Q 228 656 225 647 Q 222 640 227 632 Z',
        'M 243 466 Q 259 479 279 492 Q 385 570 407 590 Q 423 603 438 609 Q 451 613 453 621 Q 452 628 442 635 Q 417 656 392 663 Q 377 667 369 653 Q 336 569 117 420 Q 84 402 86 397 Q 86 394 93 395 Q 138 402 218 450 L 243 466 Z',
        'M 331 301 Q 352 223 319 116 Q 306 79 295 76 Q 286 72 212 82 Q 197 88 196 80 Q 195 73 206 60 Q 255 14 276 -18 Q 288 -37 303 -33 Q 322 -30 340 -5 Q 383 64 389 200 Q 380 297 363 335 Q 332 399 275 444 Q 256 460 243 466 C 217 481 195 470 218 450 Q 225 444 234 436 L 273 400 Q 288 384 300 365 L 331 301 Z',
        'M 234 436 Q 216 372 71 261 Q 56 251 75 252 Q 138 256 239 363 Q 240 366 243 367 Q 262 388 273 400 C 293 422 244 464 234 436 Z',
        'M 300 365 Q 249 239 79 104 Q 66 91 76 89 Q 97 80 165 125 Q 246 176 330 300 Q 330 301 331 301 C 349 325 312 393 300 365 Z',
        'M 582 561 Q 581 610 580 650 Q 583 669 563 676 Q 526 700 502 692 Q 484 685 498 664 Q 531 615 538 562 Q 538 558 539 552 L 543 519 Q 546 456 564 436 Q 577 424 582 436 Q 586 451 583 533 L 582 561 Z',
        'M 583 533 Q 590 537 601 540 Q 614 543 628 552 Q 634 556 632 563 Q 629 567 618 570 Q 602 567 582 561 L 539 552 Q 494 545 441 532 Q 410 531 393 528 Q 383 528 382 524 Q 381 517 387 509 Q 400 497 425 489 Q 441 482 456 487 Q 498 503 543 519 L 583 533 Z',
        'M 755 595 Q 782 677 815 734 Q 831 750 819 762 Q 806 781 768 800 Q 749 810 730 801 Q 720 794 728 782 Q 740 748 733 709 Q 715 618 668 489 Q 659 464 660 447 Q 663 432 678 441 Q 699 459 742 563 L 755 595 Z',
        'M 742 563 Q 760 560 837 567 Q 885 570 922 566 Q 935 565 940 573 Q 944 583 932 595 Q 890 629 808 606 Q 783 602 755 595 C 726 588 712 566 742 563 Z',
        'M 480 341 Q 470 348 442 353 Q 430 357 427 352 Q 420 348 428 331 Q 458 249 469 119 Q 470 79 489 52 Q 505 30 511 45 Q 514 52 515 62 L 517 96 Q 516 112 514 130 Q 496 263 492 307 C 489 336 489 336 480 341 Z',
        'M 735 84 Q 747 56 775 1 Q 784 -18 795 -16 Q 811 -12 821 15 Q 849 72 852 148 Q 858 194 862 250 Q 868 298 891 327 Q 907 346 896 360 Q 875 382 829 403 Q 810 413 778 403 Q 661 360 480 341 C 450 338 462 302 492 307 Q 495 307 600 328 L 640 336 Q 775 364 791 350 Q 800 346 799 324 Q 800 221 784 121 Q 783 103 775 102 Q 774 101 768 102 C 745 83 732 90 735 84 Z',
        'M 660 211 Q 694 218 726 222 Q 751 226 741 240 Q 729 255 703 259 Q 687 262 661 255 L 612 240 Q 575 228 543 218 Q 525 214 548 199 Q 557 192 575 196 Q 594 200 613 202 L 660 211 Z',
        'M 657 118 Q 658 167 660 211 L 661 255 Q 661 274 663 294 Q 669 315 640 336 C 616 354 592 357 600 328 Q 609 300 612 240 L 613 202 Q 613 165 615 110 C 616 80 656 88 657 118 Z',
        'M 515 62 Q 522 61 531 62 Q 597 75 735 84 C 765 86 781 87 768 102 Q 762 111 742 124 Q 724 134 657 118 L 615 110 Q 563 103 517 96 C 487 92 485 64 515 62 Z',
      ],
      medians: [
        [
          [390, 764],
          [325, 760],
          [264, 705],
          [194, 655],
          [159, 634],
          [149, 634],
          [147, 627],
        ],
        [
          [119, 581],
          [166, 532],
          [171, 508],
        ],
        [
          [236, 641],
          [268, 610],
          [283, 582],
        ],
        [
          [445, 621],
          [395, 625],
          [349, 575],
          [278, 513],
          [186, 448],
          [90, 398],
        ],
        [
          [226, 451],
          [243, 449],
          [271, 426],
          [320, 371],
          [355, 292],
          [362, 189],
          [341, 82],
          [330, 58],
          [308, 32],
          [302, 28],
          [287, 34],
          [203, 77],
        ],
        [
          [237, 430],
          [242, 399],
          [204, 352],
          [139, 295],
          [79, 259],
        ],
        [
          [301, 358],
          [300, 303],
          [225, 205],
          [146, 134],
          [81, 96],
        ],
        [
          [507, 677],
          [547, 647],
          [559, 584],
          [565, 478],
          [573, 439],
        ],
        [
          [390, 518],
          [445, 508],
          [622, 560],
        ],
        [
          [736, 791],
          [754, 780],
          [776, 747],
          [718, 560],
          [670, 447],
        ],
        [
          [746, 567],
          [779, 582],
          [857, 592],
          [896, 590],
          [930, 578],
        ],
        [
          [435, 344],
          [459, 321],
          [466, 300],
          [501, 51],
        ],
        [
          [489, 338],
          [505, 326],
          [584, 336],
          [609, 346],
          [651, 350],
          [780, 379],
          [809, 376],
          [835, 356],
          [844, 342],
          [814, 98],
          [793, 61],
          [794, -1],
        ],
        [
          [545, 208],
          [686, 238],
          [731, 233],
        ],
        [
          [608, 325],
          [635, 301],
          [636, 143],
          [621, 119],
        ],
        [
          [519, 67],
          [540, 82],
          [691, 104],
          [761, 102],
        ],
      ],
    },
  ],
  [
    '咪',
    {
      character: '咪',
      strokes: [
        'M 137 557 Q 124 564 96 570 Q 83 574 79 568 Q 72 562 80 545 Q 108 479 121 374 Q 122 343 140 321 Q 156 297 163 313 Q 169 326 169 350 L 166 384 Q 153 499 152 528 C 151 551 151 551 137 557 Z',
        'M 297 410 Q 313 506 340 535 Q 362 562 338 576 Q 272 609 271 609 Q 261 609 250 602 Q 216 580 137 557 C 108 548 123 519 152 528 Q 158 531 232 549 Q 254 555 259 546 Q 268 539 260 501 Q 253 458 243 408 C 237 379 292 380 297 410 Z',
        'M 169 350 Q 223 363 314 377 Q 324 378 325 388 Q 325 395 297 410 C 280 419 272 417 243 408 Q 200 395 166 384 C 137 375 140 343 169 350 Z',
        'M 413 591 Q 434 569 455 540 Q 468 527 483 525 Q 493 524 501 537 Q 507 550 502 582 Q 496 612 414 640 Q 399 644 392 643 Q 386 639 387 627 Q 388 615 413 591 Z',
        'M 714 668 Q 699 635 654 575 Q 644 557 662 561 Q 702 583 725 600 Q 773 640 801 653 Q 823 665 811 682 Q 796 698 769 712 Q 744 725 732 721 Q 719 720 724 706 Q 728 687 714 668 Z',
        'M 607 432 Q 683 450 827 467 Q 839 466 848 479 Q 849 492 825 503 Q 788 527 715 503 Q 618 482 607 476 L 557 466 Q 517 460 361 439 Q 330 435 354 416 Q 387 391 434 399 Q 473 408 523 416 L 607 432 Z',
        'M 606 420 Q 606 426 607 432 L 607 476 Q 607 683 625 751 Q 640 778 613 798 Q 597 810 564 829 Q 539 847 517 829 Q 511 825 518 807 Q 554 756 554 687 Q 557 506 557 466 L 556 362 Q 556 188 535 98 Q 517 71 530 32 Q 537 5 549 -21 Q 556 -40 565 -43 Q 572 -50 579 -41 Q 588 -37 599 -14 Q 609 13 608 47 Q 607 87 606 376 L 606 420 Z',
        'M 523 416 Q 447 272 248 123 Q 235 111 244 109 Q 268 100 352 153 Q 427 198 556 362 C 608 428 546 458 523 416 Z',
        'M 606 376 Q 678 270 770 154 Q 788 133 819 133 Q 927 133 960 138 Q 970 139 972 145 Q 973 151 959 158 Q 818 216 751 274 Q 682 338 606 420 C 586 442 589 401 606 376 Z',
      ],
      medians: [
        [
          [88, 560],
          [116, 534],
          [122, 520],
          [152, 319],
        ],
        [
          [152, 534],
          [169, 551],
          [259, 577],
          [275, 573],
          [301, 550],
          [275, 435],
          [249, 415],
        ],
        [
          [173, 356],
          [189, 373],
          [242, 387],
          [297, 392],
          [316, 386],
        ],
        [
          [396, 635],
          [460, 586],
          [483, 545],
        ],
        [
          [731, 712],
          [751, 692],
          [759, 674],
          [718, 624],
          [657, 568],
        ],
        [
          [356, 428],
          [424, 422],
          [765, 487],
          [837, 481],
        ],
        [
          [532, 819],
          [557, 800],
          [585, 766],
          [582, 287],
          [576, 147],
          [566, 65],
          [573, 2],
          [570, -28],
        ],
        [
          [552, 415],
          [536, 398],
          [529, 370],
          [504, 336],
          [404, 224],
          [332, 162],
          [249, 115],
        ],
        [
          [611, 413],
          [619, 383],
          [669, 322],
          [740, 242],
          [805, 182],
          [964, 145],
        ],
      ],
    },
  ],
  [
    '喵',
    {
      character: '喵',
      strokes: [
        'M 135 502 Q 123 508 97 513 Q 84 517 81 512 Q 74 506 82 490 Q 112 420 124 306 Q 125 272 144 248 Q 160 226 167 241 Q 174 262 171 302 L 167 335 Q 151 444 150 474 C 149 496 149 496 135 502 Z',
        'M 309 364 Q 322 458 346 488 Q 368 516 342 530 Q 321 540 292 556 Q 274 565 229 537 Q 187 519 135 502 C 107 493 122 463 150 474 Q 162 480 234 499 Q 258 506 264 497 Q 274 488 256 363 C 252 333 305 334 309 364 Z',
        'M 171 302 Q 214 317 323 335 Q 333 336 333 345 Q 333 352 309 364 C 290 374 285 372 256 363 Q 207 347 167 335 C 138 326 143 292 171 302 Z',
        'M 729 577 Q 792 586 903 581 Q 919 580 923 589 Q 927 599 915 614 Q 866 656 785 633 Q 766 630 746 626 L 687 616 Q 620 606 548 594 L 501 588 Q 438 581 371 573 Q 343 570 362 555 Q 408 527 478 545 Q 493 549 509 551 L 552 559 Q 576 568 678 573 L 729 577 Z',
        'M 548 594 Q 542 639 540 675 Q 541 694 521 701 Q 479 720 457 713 Q 438 703 454 684 Q 488 639 501 588 L 509 551 Q 515 520 519 488 Q 525 466 539 452 Q 552 442 557 453 Q 560 466 552 559 L 548 594 Z',
        'M 746 626 Q 752 656 779 708 Q 794 724 781 736 Q 766 754 728 770 Q 709 780 690 769 Q 680 762 689 750 Q 704 719 687 616 L 678 573 Q 659 507 661 476 Q 664 460 680 474 Q 695 486 729 577 L 746 626 Z',
        'M 420 383 Q 402 393 382 399 Q 375 402 370 396 Q 361 390 371 379 Q 417 298 422 116 Q 423 52 462 24 L 463 23 Q 481 17 483 49 Q 483 53 483 58 L 480 89 Q 459 209 454 344 C 453 366 447 369 420 383 Z',
        'M 726 79 Q 735 54 757 8 Q 764 -8 775 -8 Q 790 -5 799 19 Q 833 85 839 173 Q 846 228 853 293 Q 857 342 880 371 Q 896 392 884 405 Q 862 426 816 446 Q 797 456 764 444 Q 632 399 420 383 C 390 381 424 339 454 344 L 575 363 L 616 370 Q 770 401 789 384 Q 814 333 771 120 Q 768 101 760 98 Q 757 97 749 99 C 723 86 723 86 726 79 Z',
        'M 638 216 Q 680 223 717 227 Q 742 230 733 244 Q 723 260 695 265 Q 673 269 640 260 L 590 246 Q 547 236 508 226 Q 490 222 512 206 Q 518 202 590 209 L 638 216 Z',
        'M 633 110 Q 634 167 638 216 L 640 260 Q 640 291 643 323 Q 652 347 616 370 C 591 387 566 392 575 363 Q 587 329 590 246 L 590 209 Q 590 166 591 102 C 591 72 632 80 633 110 Z',
        'M 483 58 Q 486 58 491 58 Q 566 73 726 79 C 756 80 770 81 749 99 Q 716 132 636 111 Q 635 111 633 110 L 591 102 Q 531 95 480 89 C 450 85 453 58 483 58 Z',
      ],
      medians: [
        [
          [88, 504],
          [116, 480],
          [121, 468],
          [156, 248],
        ],
        [
          [144, 498],
          [157, 493],
          [264, 527],
          [284, 524],
          [297, 513],
          [306, 502],
          [286, 390],
          [262, 376],
        ],
        [
          [174, 308],
          [192, 326],
          [263, 344],
          [324, 344],
        ],
        [
          [365, 565],
          [385, 559],
          [450, 560],
          [590, 584],
          [829, 612],
          [886, 606],
          [911, 595],
        ],
        [
          [463, 698],
          [506, 667],
          [524, 598],
          [539, 486],
          [548, 457],
        ],
        [
          [696, 759],
          [714, 751],
          [737, 717],
          [704, 574],
          [671, 477],
        ],
        [
          [376, 388],
          [404, 366],
          [424, 332],
          [446, 127],
          [468, 29],
        ],
        [
          [430, 383],
          [439, 374],
          [474, 367],
          [626, 388],
          [780, 419],
          [800, 417],
          [824, 401],
          [836, 386],
          [820, 241],
          [795, 91],
          [775, 62],
          [775, 5],
        ],
        [
          [510, 216],
          [552, 219],
          [666, 243],
          [699, 245],
          [722, 238],
        ],
        [
          [583, 360],
          [614, 332],
          [613, 135],
          [597, 111],
        ],
        [
          [487, 66],
          [504, 77],
          [650, 96],
          [740, 97],
        ],
      ],
    },
  ],
  [
    '叫',
    {
      character: '叫',
      strokes: [
        'M 205 585 Q 192 592 164 598 Q 151 602 147 596 Q 140 590 148 573 Q 181 494 194 366 Q 195 327 216 301 Q 232 277 239 293 Q 248 315 244 357 L 240 391 Q 221 524 220 557 C 219 579 219 579 205 585 Z',
        'M 396 421 Q 414 533 448 566 Q 464 587 448 603 Q 369 654 312 621 Q 258 600 205 585 C 176 577 191 548 220 557 Q 227 558 237 563 Q 345 590 349 586 Q 361 576 360 552 Q 353 479 341 421 C 335 392 391 391 396 421 Z',
        'M 244 357 Q 388 385 410 389 Q 420 390 421 400 Q 421 407 396 421 C 376 432 370 430 341 421 Q 283 402 240 391 C 211 383 215 351 244 357 Z',
        'M 773 477 Q 671 461 595 434 Q 573 428 578 453 Q 584 529 590 607 Q 593 634 582 649 Q 564 668 537 682 Q 524 688 513 685 Q 501 679 510 659 Q 544 586 535 512 Q 528 442 500 409 Q 481 388 490 374 Q 503 358 523 346 Q 539 339 549 348 Q 576 381 706 420 Q 757 433 769 431 Q 770 432 773 432 C 803 435 803 482 773 477 Z',
        'M 771 2 Q 777 -26 785 -37 Q 791 -44 799 -42 Q 817 -33 823 34 Q 833 137 831 238 Q 827 323 831 657 Q 831 703 845 768 Q 848 787 832 798 Q 795 823 779 826 Q 764 830 746 821 Q 736 811 745 800 Q 751 787 758 776 Q 774 746 775 703 Q 776 606 773 477 L 773 432 Q 767 48 771 2 Z',
      ],
      medians: [
        [
          [156, 588],
          [184, 562],
          [191, 544],
          [228, 299],
        ],
        [
          [213, 581],
          [247, 579],
          [352, 610],
          [371, 605],
          [402, 579],
          [373, 448],
          [348, 426],
        ],
        [
          [248, 386],
          [259, 377],
          [348, 400],
          [395, 404],
          [411, 399],
        ],
        [
          [520, 673],
          [549, 640],
          [559, 616],
          [562, 555],
          [549, 404],
          [591, 407],
          [751, 452],
          [766, 470],
        ],
        [
          [755, 810],
          [776, 799],
          [803, 771],
          [806, 747],
          [795, -31],
        ],
      ],
    },
  ],
]);
