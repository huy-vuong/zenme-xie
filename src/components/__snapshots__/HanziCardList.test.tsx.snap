// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing given empty text 1`] = `<DocumentFragment />`;
